File: Code/Ifly/Presentation.cs
using System.Collections.Generic;

namespace Ifly
{
    /// <summary>
    /// Represents presentation chart provider type.
    /// </summary>
    public enum PresentationChartProviderType
    {
        /// <summary>
        /// Chart.js.
        /// </summary>
        ChartJS = 0,

        /// <summary>
        /// Google Charts.
        /// </summary>
        GoogleCharts = 1
    }

    /// <summary>
    /// Represents presentation description mode.
    /// </summary>
    public enum PresentationSlideDescriptionType
    {
        /// <summary>
        /// Always use slide description.
        /// </summary>
        Always = 0,

        /// <summary>
        /// Don't use slide description.
        /// </summary>
        Never = 1
    }

    /// <summary>
    /// Represents a presentation. This class cannot be inherited.
    /// </summary>
    public sealed class Presentation : Storage.Record
    {
        /// <summary>
        /// Gets or sets the Id of the user who has created this presentation.
        /// </summary>
        public int UserId { get; set; }

        /// <summary>
        /// Gets or sets value indicating whether presentation is active.
        /// </summary>
        public bool IsActive { get; set; }

        /// <summary>
        /// Gets or sets value indicating whether presentation is archived.
        /// </summary>
        public bool IsArchived { get; set; }

        /// <summary>
        /// Gets or sets the total number of impressions for this presentation.
        /// </summary>
        public int TotalImpressions { get; set; }

        /// <summary>
        /// Gets or sets the charts type.
        /// </summary>
        public PresentationChartProviderType UseCharts { get; set; }

        /// <summary>
        /// Gets or sets slide description type.
        /// </summary>
        public PresentationSlideDescriptionType UseSlideDescription { get; set; }

        /// <summary>
        /// Gets or sets the user subscription type stamped when the presentation was last time updated.
        /// </summary>
        public SubscriptionType UserSubscriptionType { get; set; }

        /// <summary>
        /// Gets or sets the date and time when this presentation was created.
        /// </summary>
        public System.DateTime Created { get; set; }

        /// <summary>
        /// Gets or sets the presentation title.
        /// </summary>
        public string Title { get; set; }

        /// <summary>
        /// Gets or sets the name of the theme that is applied to this presentation.
        /// </summary>
        public string Theme { get; set; }

        /// <summary>
        /// Gets or sets the background image.
        /// </summary>
        public string BackgroundImage { get; set; }

        /// <summary>
        /// Gets or sets the list of slides that are part of a given presentation.
        /// </summary>
        public IList<Slide> Slides { get; set; }

        /// <summary>
        /// Gets or sets the publish settings.
        /// </summary>
        public PublishConfiguration PublishSettings { get; set; }

        /// <summary>
        /// Gets or sets the integration settings.
        /// </summary>
        public IntegrationSettings IntegrationSettings { get; set; }

        /// <summary>
        /// Gets or sets the presenter settings.
        /// </summary>
        public PresenterModeConfiguration PresenterSettings { get; set; }

        /// <summary>
        /// Initializes a new instance of an object.
        /// </summary>
        public Presentation()
        {
            this.Theme = "clean";
            this.IsActive = true;
            this.IsArchived = false;
            this.Slides = new List<Slide>();
            this.Created = System.DateTime.UtcNow;
            this.PublishSettings = new PublishConfiguration();
            this.IntegrationSettings = new IntegrationSettings();
            this.PresenterSettings = new PresenterModeConfiguration();
        }
    }
}

File: Code/Ifly.Web.Editor/Assets/js/Models/Presentation.ts
/// <reference path="../Typings/knockout.d.ts" />
/// <reference path="Slide.ts" />
/// <reference path="IModel.ts" />

module Ifly.Models {
    /** Represents a presentation (infographic). */
    export class Presentation implements IModel {
        /** Gets or sets the presentation Id. */
        public id: KnockoutObservable<number>;

        /** Gets or sets user Id. */
        public userId: KnockoutObservable<number>;

        /** Gets or sets the presentation title. */
        public title: KnockoutObservable<string>;

        /** Gets or sets the date and time when presentation was created. */
        public created: KnockoutObservable<Date>;

        /** Gets or sets the presentation theme. */
        public theme: KnockoutObservable<string>;

        /** Gets or sets the background image. */
        public backgroundImage: KnockoutObservable<string>;

        /** Gets or sets value indicating whether presentation is active. */
        public isActive: KnockoutObservable<boolean>;

        /** Gets or sets value indicating whether presentation is archived. */
        public isArchived: KnockoutObservable<boolean>;

        /** Gets or sets the total number of impressions for this presentation. */
        public totalImpressions: KnockoutObservable<number>;

        /** Gets or sets the chart type. */
        public useCharts: KnockoutObservable<PresentationChartProviderType>;

        /** Gets or sets the slide description type. */
        public useSlideDescription: KnockoutObservable<PresentationSlideDescriptionType>;

        /** Gets or sets the list of slides that are part of a given presentation. */
        public slides: KnockoutObservableArray<Slide>;

        /** Gets or sets the publish settings. */
        public publishSettings: KnockoutObservable<PublishConfiguration>;

        /** Gets or sets integration settings. */
        public integrationSettings: KnockoutObservable<IntegrationSettings>;

        /** Gets or sets presenter settings. */
        public presenterSettings: KnockoutObservable<PresenterModeConfiguration>;

        /** Gets or sets the selected slide index. */
        public selectedSlideIndex: KnockoutObservable<number>;

        /** Gets or sets value indicating whether slide is being selected. */
        public isSelectingSlide: KnockoutObservable<boolean>;

        /** Gets or sets the selected slide. */
        public selectedSlide: KnockoutComputed<Slide>;

        /** 
         * Initializes a new instance of an object.
         * @param {object} data Data to load from.
         */
        constructor(data?: any) {
            this.id = ko.observable<number>();
            this.userId = ko.observable<number>();
            this.title = ko.observable<string>();
            this.theme = ko.observable<string>();
            this.created = ko.observable<Date>();
            this.backgroundImage = ko.observable<string>();
            this.isActive = ko.observable<boolean>();
            this.isArchived = ko.observable<boolean>();
            this.totalImpressions = ko.observable<number>();
            this.useCharts = ko.observable<PresentationChartProviderType>(PresentationChartProviderType.googleCharts);
            this.useSlideDescription = ko.observable<PresentationSlideDescriptionType>(PresentationSlideDescriptionType.never);
            this.slides = ko.observableArray<Slide>();
            this.publishSettings = ko.observable<PublishConfiguration>(new PublishConfiguration());
            this.integrationSettings = ko.observable<IntegrationSettings>(new IntegrationSettings());
            this.presenterSettings = ko.observable<PresenterModeConfiguration>(new PresenterModeConfiguration());

            this.isSelectingSlide = ko.observable<boolean>();

            this.selectedSlideIndex = ko.observable(-1);
            this.selectedSlide = ko.computed<Slide>(() => {
                var i = this.selectedSlideIndex();
                return i >= 0 && i < this.slides().length ? this.slides()[i] : null;
            });

            this.load(data);
        }

        /** 
         * Populates object state from the given data.
         * @param {object} data Data to load from.
         */
        public load(data: any) {
            var slides = [], publishSettings = null;

            data = data || {};

            this.id(data.id || data.Id || 0);
            this.userId(data.userId || data.UserId || 0);
            this.title(data.title || data.Title || '');

            if (!data.created && !data.Created) {
                this.created(new Date());
            } else {
                this.created(moment(data.created || data.Created).toDate());
            }
            
            this.theme(data.theme || data.Theme || 'clean');
            this.backgroundImage(data.backgroundImage || data.BackgroundImage || '');
            this.isActive(data.isActive || data.IsActive);
            this.isArchived(data.isArchived || data.IsArchived);
            this.totalImpressions(data.totalImpressions || data.TotalImpressions || 0);
            this.useCharts(data.useCharts || data.UseCharts || 0);
            this.useSlideDescription(data.useSlideDescription || data.UseSlideDescription || 0);

            this.slides.removeAll();

            slides = data.slides || data.Slides;

            if (slides && slides.length) {
                slides.sort((x, y) => (x.order || x.Order || 0) - (y.order || y.Order || 0));

                ko.utils.arrayForEach(slides, (s) => {
                    this.slides.push(new Slide(s));
                });
            }

            this.publishSettings().load(data.publishSettings || data.PublishSettings);
            this.integrationSettings().load(data.integrationSettings || data.IntegrationSettings);
            this.presenterSettings().load(data.presenterSettings || data.PresenterSettings);
        }

        /** Serializes object state. */
        public serialize() {
            return {
                id: this.id(),
                userId: this.userId(),
                title: this.title(),
                created: this.created(),
                theme: this.theme(),
                backgroundImage: this.backgroundImage(),
                isActive: this.isActive(),
                isArchived: this.isArchived(),
                totalImpressions: this.totalImpressions(),
                useCharts: this.useCharts(),
                useSlideDescription: this.useSlideDescription(),
                slides: ko.utils.arrayMap(this.slides(), (s) => {
                    return (<Slide>s).serialize();
                }),
                publishSettings: this.publishSettings().serialize(),
                integrationSettings: this.integrationSettings().serialize(),
                presenterSettings: this.presenterSettings().serialize()
            };
        }
    }
}